### PR TITLE
Sparkler Configuration Update

### PR DESCRIPTION
**Pull request by SHASHANK-PRO-05**
I have added validations and type check in sparkler configuration. Also, I have added some TODOs comment which I think should be given a thought to. @thammegowda @karanjeets do give it a look. Let me know any suggestions, will do the needful. Also, I found a very interesting thing, there are multiple sources of the same type of files present.
```
./resources/sparkler-default.yaml
./sparkler-app/target/test-classes/sparkler-default.yaml
./sparkler-app/src/test/resources/sparkler-default.yaml
./conf/sparkler-default.yaml
./sparkler-api/target/classes/sparkler-default.yaml
```
There are multiple files like sparkler-default.yaml. #93

### CODE REVIEW

**Comment by thammegowda** [comment]
@SHASHANK-PRO-05 Awesome.  This is a huge PR! Thanks for doing this.
 we will take a day or two to go through all the changes and test them.

**Comment by SHASHANK-PRO-05** [comment]
Sure, Thanks.

**Comment by SHASHANK-PRO-05** [comment]
Not able to understand the issue with Travis. Can anyone explain it?

**Comment by thammegowda** [comment]
@SHASHANK-PRO-05 Good work.

> There are multiple files like sparkler-default.yaml.

That's right. this is as per the maven conventions. Some are simply copies made by maven. [More here](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html)

+ files under `**target/ directory` is generated by maven for packaging… ignore them. when you do `mvn clean` they will disappear.
+ files under `src/test/resources` are meant to be used by test cases. each maven module can have its own test resources.
+ files under `./conf/` and `src/main/resources` is meant to be used for packaging and release. all these will be tracked in git.

+ files under `./resources` are a copy of `./conf`  these could be modified by users and these are not tracked by git.

So in summary,  test cases are using the files in `src/test/resources`.

**Comment by thammegowda** [comment]
> Not able to understand the issue with Travis. Can anyone explain it?

That was an invalid configuration on Travis, no need to worry. We are working on it!

 @buggtb why Travis treating this PR as ruby build
```
{
  "language": "ruby",
  "group": "stable",
  "dist": "precise",
  "os": "linux"
}
```

**Comment by thammegowda** [comment]
@SHASHANK-PRO-05 
+1 for using hibernate validations
+1 for writing extensive tests

However, refer to the code in the constructor for `SparklerCofig`: 
 https://github.com/USCDataScience/sparkler/pull/95/files#diff-10d40680d9c04ee4cb933137e6ef1e5dR207 

I spent many hours to go through this PR and thus I can understand that you too have spent lots of time to work on this PR.
I feel the parsing code {c/s}ould be improved:
- It is too much (verbose) code. For every property, checking if it exists, reading, setting to field, validating.... Thats something we need to hide to make it easy to maintain.  __Rule of thumb: Write absolute minimum code that is required to get the stuff done. Less code means less worries __ 
- It could be challending when we need to add more fields.
- Deserialization should often be coupled with serialization. Yaml --> Java --> Yaml should be possible. So again too much code to write Yaml serialization code.
 
I totally appreciate your efforts, please dont be disappointed by this feedback.  There were many good parts in your code I used many of those pieces and created an example. Please review and share your ideas. 

CC @karanjeets  feel free to jump in and correct us!

## Config
```yaml
!sparkler
crawldb:
  uri: http://localhost:8983/solr/crawldb

fetcher:
  headers:
    Name: Value
  serverDelay: 1000
  userAgents:
    - UA 1
    - UA 2
    - UA 3
someFile: pom.xml
```

## Code 
```java

interface BaseConfig {}

public class SparklerConfig implements BaseConfig {

    private static final Logger LOG = LoggerFactory.getLogger(SparklerConfig.class);

    ////// Nested classes --can be moved to outside in a same package
    public static class CrawldbConfig implements BaseConfig {

        @NotNull
        private URI uri;

        public URI getUri() {
            return uri;
        }

        public void setUri(URI uri) {
            this.uri = uri;
        }

        @Override
        public String toString() {
            return "CrawldbConfig{" +
                    "uri=" + uri +
                    '}';
        }
    }

    public static class FetcherConfig implements BaseConfig{

        @Range(min = 1, max = 60*60*1000)
        private long serverDelay;

        private Map<String, String> headers;

        private List<String> userAgents;

        public long getServerDelay() {
            return serverDelay;
        }

        public void setServerDelay(long serverDelay) {
            this.serverDelay = serverDelay;
        }

        public Map<String, String> getHeaders() {
            return headers;
        }

        public void setHeaders(Map<String, String> headers) {
            this.headers = headers;
        }

        public List<String> getUserAgents() {
            return userAgents;
        }

        public void setUserAgents(List<String> userAgents) {
            this.userAgents = userAgents;
        }

        @Override
        public String toString() {
            return "FetcherConfig{" +
                    "serverDelay=" + serverDelay +
                    ", headers=" + headers +
                    ", userAgents=" + userAgents +
                    '}';
        }
    }
    //// This class applies any customizations on
    private static class SparklerConfigRepresenter extends Representer {
        SparklerConfigRepresenter(){
            super();
            this.addClassTag(SparklerConfig.class, new Tag("!sparkler"));
        }

        @Override
        protected NodeTuple representJavaBeanProperty(
                Object javaBean, Property property,
                Object propertyValue, Tag customTag) {
            //NOTE: snakeyaml v1.18 doesnt serialize types like File and URI properly to string, instead
            //     an empty dictionary is inserted like '{}' at the value, so this fix is applied
            // TODO: send this as PR to snake YAML
            NodeTuple tuple = super.representJavaBeanProperty(javaBean, property,
                    propertyValue, customTag);
            // This property can be serialized to string
            if (propertyValue != null
                    && STRING_SERIALIZABLE.contains(property.getType())){
                // but it has been converted to object/bean of properties
                if (Object.class.equals(tuple.getValueNode().getType())){
                    // and mapper found zero pairs of properties
                    if (tuple.getValueNode().getClass().isAssignableFrom(MappingNode.class)
                            && ((MappingNode) tuple.getValueNode()).getValue().isEmpty()) {
                        Node value = representScalar(tuple.getKeyNode().getTag(), propertyValue.toString());
                        tuple = new NodeTuple(tuple.getKeyNode(), value);
                    }
                }
            }
            return tuple;
        }
    }

    private static final Validator VALIDATOR =
            Validation.buildDefaultValidatorFactory().getValidator();
    private static final Yaml YAML;
    private static final Set<Class<?>> STRING_SERIALIZABLE = new HashSet<>(Arrays.asList(
            URI.class, File.class, Path.class, URL.class));
    static {
        Constructor constructor = new Constructor();
        constructor.addTypeDescription(new TypeDescription(SparklerConfig.class, "!sparkler"));
        YAML = new Yaml(constructor, new SparklerConfigRepresenter());
    }

    // Members
    @Valid
    private CrawldbConfig crawldb;

    @Valid
    private FetcherConfig fetcher;

    private File someFile;

    public File getSomeFile() {
        return someFile;
    }

    public void setSomeFile(File someFile) {
        this.someFile = someFile;
    }

    public CrawldbConfig getCrawldb() {
        return crawldb;
    }

    public void setCrawldb(CrawldbConfig crawldb) {
        this.crawldb = crawldb;
    }

    public FetcherConfig getFetcher() {
        return fetcher;
    }

    public void setFetcher(FetcherConfig fetcher) {
        this.fetcher = fetcher;
    }

    @Override
    public String toString() {
        return "SparklerConfig{" +
                "crawldb=" + crawldb +
                ", fetcher=" + fetcher +
                '}';
    }

    public static SparklerConfig load(InputStream stream)
            throws SparklerConfigException {

        synchronized (SparklerConfig.class){
            SparklerConfig config = (SparklerConfig) YAML.load(stream);
            //validate
            Set<ConstraintViolation<BaseConfig>> violations = VALIDATOR.validate(config);
            if (!violations.isEmpty()){
                throw new SparklerConfigException("Found " +
                        violations.size() +" violations", violations);
            }
            return config;
        }

    }

    public String toYamlString(){
        return YAML.dump(this);
    }

    /**
     *  This exception is thrown when the config is invalid
     */
    public static class SparklerConfigException extends SparklerException {

        private Set<ConstraintViolation<BaseConfig>> violations;
        public SparklerConfigException(String message,
                                       Set<ConstraintViolation<BaseConfig>> violations) {
            super(message);
            this.violations = violations;
        }

        public Set<ConstraintViolation<BaseConfig>> getViolations() {
            return violations;
        }
    }

    public static void main(String[] args)
            throws FileNotFoundException, SparklerException {
        FileInputStream in = new FileInputStream("sparkler-conf-new.yaml");
        try {
            SparklerConfig config = SparklerConfig.load(in);
            System.out.println(config);
            System.out.println("==YAML DUMP==");
            String yamlDump = config.toYamlString();
            System.out.println(yamlDump);
            // test serialization and deserialization
            SparklerConfig config2 = SparklerConfig.load(
                    new ByteArrayInputStream(yamlDump.getBytes()));
            System.out.println("Config2::" + config2);
            assert config.toString().equals(config2.toString());

        } catch (SparklerConfigException e){
            e.printStackTrace();
            System.out.println(e.getViolations());
        }
    }
}

```

**Comment by SHASHANK-PRO-05** [comment]
@thammegowda
Points Taken
1) There should be easy serialization from YAML<->Java Bean.
2) Less Code, less headache. I was doing that because I though it should fail bombastically with proper errors. No issues.

> I totally appreciate your efforts, please don't be disappointed by this feedback
Thanks, and I am not disappointed :) . The best should be merged with master.

Cool, will improvise the PR on the feedback.

**Comment by buggtb** [comment]
@thammegowda Travis builds on existing forks will fail because they don't have a .travis.yml config file, future forks and branches should build fine because they will pick up this file.

**Comment by SHASHANK-PRO-05** [comment]
@buggtb got the .travis.yaml file in the pull. Thanks.
 @thammegowda thanks for this example. This makes a lot of things clear to me. 

But I think we should not do this implementation in SparklerConfiguration, now. It is dependent on JSONObject Class. I would propose that I move this implementation to a different class independent of JSONOBject class in which we can directly use getters and setters. Also, this way we will be able to move things easily step by step as the previous code will not be touched.

**Comment by thammegowda** [comment]
> But I think we should not do this implementation in SparklerConfiguration, now. It is dependent on JSONObject Class. I would propose that I move this implementation to a different class independent of JSONOBject class in which we can directly use getters and setters. Also, this way we will be able to move things easily step by step as the previous code will not be touched.

👍

**Comment by SHASHANK-PRO-05** [comment]
@thammegowda @karanjeets 
I am still working on this but can you guys have a look at the configuration I am creating 
[here](https://github.com/SHASHANK-PRO-05/sparkler/blob/dev/conf/sparklerConfigurations-default.yaml)? The previous style of configuration is a little cumbersome to parse with SnakeYaml. Creating a custom map to parse the keys of a JavaBean would not be good in a long run. Rather, SnakeYaml should do that job for us. Let me know your suggestions, I will proceed the development after that. Also, if you get time, have a look into the new config (JavaBean)[https://github.com/SHASHANK-PRO-05/sparkler/blob/dev/sparkler-api/src/main/java/edu/usc/irds/sparkler/SparklerConfig.java]. Thanks

**Comment by thammegowda** [comment]
@SHASHANK-PRO-05 Sorry, I could not review it in detail so far. I will do it this weekend. 

In a quick review, I see you are going in the right direction of delegating the parsing task to SnakeYaml to parse as much as possible.

Quick feedback:
+ prefer shorter names. I think adding `properties` suffix to all top level keys is redundant (since we are in config we know they are all config). Find the right balance between shortnames that makes no sense and the long names that have more chances of typos! I leave this to you to decide.

+ We may need to parse config in multi-stages. The first pass parses the config at the root level. The second pass shall parse config of individual plugins. Reason: we cannot declare all the config objects of plugins in the sparkler-API module. We have to let plugins to declare its custom config blocks. I don't have a clear plan as of now, I will share it with you in few days. As always, you are encouraged to make a plan too.  Then we can have a discussion
